cirq/google/calibration/workflow.py

```diff
@@ -337,19 +351,150 @@ def run_characterizations(
+
+            if isinstance(op.gate, (MeasurementGate, SingleQubitGate)):
+                other.append(op)
+            else:
+                a, b = op.qubits
+
+                translated_gate = gates_translator(op.gate)
+                if translated_gate is None:
+                    raise IncompatibleMomentError(
+                        f'Moment {moment} contains unsupported non-single qubit operation {op}'
+                    )
+
+                if parameters is None:
+                    raise ValueError(f'Missing characterization data for moment {moment}')
+
+                pair_parameters = parameters.get_parameters(a, b)
+                if pair_parameters is None:
+                    raise ValueError(
+                        f'Missing characterization data for pair {(a, b)} in {parameters}'
+                    )
+                pair_parameters = pair_parameters.merge_with(default_phases)
+
+                corrected = PhaseCorrectedFSimOperations(
+                    (a, b), translated_gate, pair_parameters, characterization_index
+                )
+                decompositions.append(corrected.operations)
+
+                if new_moment_mapping is None:
+                    new_moment_mapping = corrected.moment_allocations
+                else:
+                    assert (
+                        new_moment_mapping == corrected.moment_allocations
+                    ), f'Inconsistent decompositions with a moment {moment}'
+
+        if other and decompositions:
+            raise IncompatibleMomentError(f'Moment {moment} contains mixed operations')
+        elif other:
+            compensated += Moment(other)
+            compensated_allocations.append(characterization_index)
+        elif decompositions:
+            for operations in zip_longest(*decompositions, fillvalue=()):
+                compensated += Moment(operations)
+            assert new_moment_mapping is not None  # Required for mypy
+            compensated_allocations += new_moment_mapping
+
+    return CircuitCalibration(compensated, compensated_allocations)
+
+
+class PhaseCorrectedFSimOperations:
+    """Operations that compensate for zeta, chi and gamma angles of an approximate FSimGate gate.
 
     Attributes:
-        results: List of PhasedFSimCalibrationResult for each characterized moment.
-        moment_allocations: List of indices to element in results list for each moment in the
-            supplied circuit. If None occurs at certain position, it means that the related moment
-            does not require characterization.
+        operations: Tuple of tuple of operations that describe the gate. The first index iterates
+            over moments of the composed operation.
+        moment_allocations: List of indices pointing to the characterizations for each moment in the
+            composed operation.
     """
 
-    results: List[PhasedFSimCalibrationResult]
-    moment_allocations: List[Optional[int]]
+    def __init__(
```

Comment:
my knee-jerk reaction is that this should be a free function that returns a "dumb" dataclass container, but maybe that's excessive. 

As a piece of evidence: when I was reading through the above function I couldn't intuit what PhaseCorrectedFsimOperations was since I expected it to be a data container but it's actually doing the "work" of figuring out the corrections (in the constructor, no less)

Reply:
You're right, this is a much better design.